File: src/temperature_monitor.cpp
#include "ros/ros.h"
#include "sensor_msgs/Temperature.h"

ros::Publisher average_pub;
float average;
float variance;


void temp_callback(const sensor_msgs::TemperatureConstPtr& temp) {
    ROS_INFO("Temperature Received: %f", temp->temperature);
    average = (average + temp->temperature) / 2; //Floading Avarage
    variance += temp->variance; //Highest Variance

    sensor_msgs::Temperature avTemp;
    avTemp.header.stamp = ros::Time::now();
    avTemp.temperature = average;
    avTemp.variance = variance;

    average_pub.publish(avTemp);
}

int main(int argc, char **argv){
    ros::init(argc, argv, "temperature_monitor_cpp", ros::init_options::AnonymousName);
    ROS_INFO("Starting temperature_monitor_cpp...");
    ros::NodeHandle nodeHandle;
    average = 0;
    variance = 0;
    average_pub = nodeHandle.advertise<sensor_msgs::Temperature>("franek/temp_average", 10);
    ros::Subscriber tempSub = nodeHandle.subscribe("franek/temperature", 10, temp_callback);

    ROS_INFO("Started temperature_monitore_cpp successfully. Listening...");

    ros::spin();   
}

File: scripts/temperature_monitor.py
#! /usr/bin/env python
import rospy, std_msgs.msg
from sensor_msgs.msg import Temperature

pub = rospy.Publisher('henri/temp_average', Temperature, queue_size=10)
average = 0
variance = 0

def callback(data):
    global average, variance, pub
    rospy.loginfo('Temperature Received: %f', data.temperature)
    average = (average + data.temperature)/2
    variance = (variance + data.variance)/2

    t = Temperature()

    h = std_msgs.msg.Header()
    h.stamp = rospy.Time.now()
    
    t.header = h
    t.temperature = average
    t.variance = variance
    
    pub.publish(t)

def listen_temp():
    rospy.init_node('temperature_monitor_py', anonymous=True)
    rospy.Subscriber('henri/temperature', Temperature, callback)
    rospy.spin()

if __name__ == '__main__':
    listen_temp()
